ppl/zqd/recruiter/recruiter.go

```diff
@@ -0,0 +1,191 @@
+
+// WorkerPool is the internal state of the recruiter system
+// The pools are exported only for use in unit tests
+type WorkerPool struct {
+	mu           sync.Mutex                // one lock for all three maps
+	freePool     map[string]WorkerDetail   // Map of all free workers
+	nodePool     map[string][]WorkerDetail // Map of nodes of slices of free workers
+	reservedPool map[string]WorkerDetail   // Map of busy workers
+	SkipSpread   bool                      // option to test algorithm performance
+}
+
+type WorkerDetail struct {
+	Addr     string
+	NodeName string
+}
+
+func NewWorkerPool() *WorkerPool {
+	rand.Seed(time.Now().UnixNano()) // for shuffling nodePool keys
+	return &WorkerPool{
+		freePool:     make(map[string]WorkerDetail),
+		nodePool:     make(map[string][]WorkerDetail),
+		reservedPool: make(map[string]WorkerDetail),
+	}
+}
+
+func (pool *WorkerPool) Register(addr string, nodename string) error {
+	if _, _, err := net.SplitHostPort(addr); err != nil {
+		return fmt.Errorf("Invalid address for Register: %w", err)
+	}
+	if nodename == "" {
+		return fmt.Errorf("Node name required for Register")
+	}
+	wd := WorkerDetail{Addr: addr, NodeName: nodename}
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	_, prs := pool.reservedPool[addr]
+	if prs {
+		return nil // ignore register for existing workers
+	}
+	pool.freePool[addr] = wd
+	_, prs = pool.nodePool[nodename]
+	if !prs {
+		pool.nodePool[nodename] = make([]WorkerDetail, 0)
+	}
+	pool.nodePool[nodename] = append(pool.nodePool[nodename], wd)
+	return nil
+}
+
+// removeFromNodePool is internal and the calling function must be holding the mutex
+func (pool *WorkerPool) removeFromNodePool(wd WorkerDetail) {
+	s := pool.nodePool[wd.NodeName]
+	if len(s) == 1 {
+		if s[0] == wd {
+			// Remove empty list from the hash so len(pool.nodePool)
+			// is a a count of nodes with available workers.
+			delete(pool.nodePool, wd.NodeName)
+		}
+		return
+	}
+	i := -1
+	for j, v := range s {
+		if v == wd {
+			i = j
+			break
+		}
+	}
+	if i == -1 {
+		panic(fmt.Errorf("expected WorkerDetail not in list: %v", wd.Addr))
+	}
+	// Overwrite removed node and truncate slice
+	s[i] = s[len(s)-1]
+	pool.nodePool[wd.NodeName] = s[:len(s)-1]
+}
+
+func (pool *WorkerPool) Deregister(addr string) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	wd, prs := pool.freePool[addr]
+	if prs {
+		pool.removeFromNodePool(wd)
+		delete(pool.freePool, addr)
+	}
+}
+
+// Unreserve removes from the reserved pool, but does not reregister.
+// The worker process should initiate register.
+func (pool *WorkerPool) Unreserve(addr string) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	delete(pool.reservedPool, addr)
+}
+
+// Recruit attempts to spread the workers across nodes.
+func (pool *WorkerPool) Recruit(n int) ([]WorkerDetail, error) {
+	if n < 1 {
+		return nil, fmt.Errorf("Recruit must request one or more workers: n=%d", n)
+	}
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	recruits := make([]WorkerDetail, 0)
```

Comment:
It's ok to return a nil slice, and you don't need to make this before usage, so this can be:

```suggestion
	var recruits []WorkerDetail
```